### ISSUE TITLE
LCU Client's new Item Set Creator doesn't recognize sets made with website

### ISSUE THREAD

**Issue by chaoticnote** [Cadrach/lol_ise#46]
It just doesn't. I tried deleting the files "ItemSets.json" and "PersistedSettings.json" while replacing the "Champions" folder with the new one. Logging onto League just replaces files to old ones I had in the Legacy Client. (A whole month's worth of updating my item sets with ISE flushed down the drain.)

**Comment by Cadrach** [comment]
I have not updated my client yet, I will take a look tonight. I will have to assess, if they changed completely the way item sets are stored it might take a few days for me to solve. I just hope they do not prevent it now...

**Comment by itorres008** [comment]
Yes, seems client pulled the old item sets it had stored I guess on their server from the old legacy client.

And what's with the duplicating the item sets under each champion? It seems like bad programming practice. Data normalization, avoiding duplicate data?

Although some of the champion foders have one item set less and some have different file dates than the rest. Will wait to see if you figure something out before experimenting. :-)

The sets are not even in alphabetical order nor can you order them manually like they allow now with masteries.

**Comment by Cadrach** [comment]
Did not have enough time yesterday (children :)) will take the time this weekend hopefully!

**Comment by Cadrach** [comment]
OK, I was able to check things, and I think I understand what is going on, but it is strange:
- The game still loads correctly the sets I created with the website (I launched a custom game and had a set "MyBrand!")
- The ingame item set editor seems to be loading sets from the ItemSets.json file.
- I added a set with the new Item Set editor, and it appeared in my next custom game, but my previous set also were there!

So my conclusion is that: the ingame Item Set Editor works on the ItemSets.json file, but the game merges both ItemSets.json + Champion & Global folder to create the sets.

Maybe what I will do next, is ask only for the ItemSets.json file for upload, and export this file only?

**Comment by chaoticnote** [comment]
Actually I sort of figured out a fix? It seems to replace files that have "Riot_Item" in the name. I tried renaming files so they wouldn't get replaced, but of course the Item Sets on serverside still come back. So in the end, I just went on the client and deleted all the old builds I had on there. Now my builds made on ISE don't get replaced.

**Comment by Cadrach** [comment]
Ok, I will probably try and generate the ItemSets.json and see if it is enough in the future, will let you know!

**Comment by itorres008** [comment]
Their data store "structure" is weird. They duplicate all the item sets on a folder per champion which is redundant and wasteful fi they can track items sets recommend for a particular champion in a more efficient way, but then the sets in the folders are not all the same. They have different modified on dates and some actually have one more or less sets than others.

Actually, I have never defined an Item set for a particular champion. They are all Global. After legacy client was retired i started using ISE, I did away with the Champions folder and copied sets to Global only and everything worked well. As matter of fact, I didn't have a Global folder, just Champions. Don't know why. Must be some sort of transition in data store scheme between Legacy and new client.

I tell just in case it helps in any way.

**Comment by Cadrach** [comment]
> Edit: I just checked and the folders under \Champion\Champion-name\Recommended are all empty now - even though they left the folders there.. Either they are doing away with this structure or it is because I don't have sets assgined to a particular champion.

I think it is the later, I still have them on my side, with 1 file per sets in each champion folder.

The issue I think I will have is the size of the file. Some people, believe it or not, have thousands of custom sets, and storing them all in one file will make it quite big. I think that is why Riot has a current limit of 20 item sets in the client. I will test soon, but it is possible that if I save the sets to this file too, I would have to stick to this limit. If so, I will probably leave the system as is, so that it provides an additionnal feature to the users, compared to the client.

**Comment by itorres008** [comment]
Thanks for the info about the folders.

About the file size, I have 17 Item Sets each with a few (6-10?) builds each. My ItemSets.json is 31KB. So, all numbers approximated, if I had 1700 it might be 3.1MB. At a modest 4 Mbps DSL (12% overhead) in one minute one can download 26MB. So, that 3,1Mb file with 1700 Sets could take 7-8 secs to download.  So you could allow even 1,000. And, most people have faster internet than 4 Mbps.

**Comment by Cadrach** [comment]
It is not so much the size of the file itself, but the loading. When the client launches the game, they would have to load the full file instead of the very small footprint of the single file per set. That would be a sound reason I think. In any case, I will try and generate a big one and see how it goes :)

**Comment by itorres008** [comment]
I don't know if you have had time to work on this, but this is what I have found - that you may already know. 
I believe if you do generate an ItemSets.json that includes all the item sets in \Config\Global\Recommended and for \Champions (which I don't use), copying ItemSets.json and the \Global and \Champions will work and solve our problem
.
My tests showed me
- \config\ItemSets.json – stores all item sets for Client Editor Tool. Updating and Saving additionally updates all sets on \config\Global\Recommended
- \config\Global\Recommended – holds individual Item Sets for Store Tool. If this folder is empty the Item Sets will not appear on the store (I presume this is legacy method and will disappear if they work on using just ItemSets.json)

My test:
I deleted ItemSets from \config\Global\Recommended, left ItemSets.json untouched.
- Item Editor reads sets fine from itemsets.json
- StoreTool in game doesn't show any item sets (pointing to it is reading sets from \config\global which I deleted)
- Exit from LoL client and \config\global remains empty (no update on exit)
- Run LoL Client again, go to Item Set Editor, modify on Item Set and hit Save, all Item Sets are saved to \config\global\recommended. (So, when new Item Set creator went live, Riot restored the last ItemSet.json file we saved on Legacy Client and wrote over \Global and \Champions wiping sets we created with ISE)

I'm thinking of manually copying and pasting my ISE sets into a ItemsSets.json file unless you think you're close to implementing it.

I have some trepidation, since the Riot files are stripped of CR, LF, and other formatting and the tools I have to look at the files reformat and beautify them and I'm not sure if that will cause problems later when Riot reads them.
(I can use Visual Studio, Notepad++ with JSON viewer plugin or an online JSON editor)

**Comment by Cadrach** [comment]
I am working on something, but it is not so simple:

- ItemsSets.json gets replaced everytime you launch the client by a server-saved version. 
- In order to not get replaced by the client, the ItemsSets.json file must contain the account id of the current account, and a timestamp. It then overwrite the server-saved version.
- So basically, I must implement a way for the user to get their account id, by typing summoner name + played on server. Once I have the account id, I will be able to generate the ItemsSets.json file.

Once this is done, I will test generating an ItemsSets.json with more than 20+ sets, which seems to be a limit currently on the client, and see how it behaves.

**Comment by itorres008** [comment]
> ItemsSets.json gets replaced everytime you launch the client by a server-saved version.

It always gets replaced, doesn't Riot check if the local file is different, the same or timestamp to avoid download?
If it doesn't, then if the user copies the ISE files just before Riot saves to the server that woul solve the problem. It would save the file to server either on Client exit (which I doubt) or when User hits Save in the Item Set Editor. So, I believe we can solve the problem if we:
- start LoL client - server copy is copied to \config as you suspect
- we copy ISE file to \config - overwriting old server copy
- do some minimal edit to one set on Item Set Editor and hit save - this would save our new ISE ItemSet.json to server
I'l try to test this hioithesis. :-)

> So basically, I must implement a way for the user to get their account id, by typing summoner name + played on server. Once I have the account id, I will be able to generate the ItemsSets.json file.

I think you can save that effort, since you have the account ID in the original ItemSets.json the user uploads to the ISE tool.

**Comment by Cadrach** [comment]
It will not work this way, ItemSet.json gets load only at client launch, if you replace it you must restart the client to load it, and doing so resets it :)

I have the account id that is true, but what of people that already have sets in the tool and want to redownload them, or people with multiple account per machine (family compute for example)

The way I will build it should cover all these edge case and be rather simple to use (applying sets to the game should not be complicated)

**Comment by itorres008** [comment]
> I have the account id that is true, but what of people that already have sets in the tool and want to redownload them, or people with multiple account per machine (family compute for example)

Yes, I see. The multiple user is a thing to handle. And the ISE currently doesn't read and store the original account number. You could have them upload the IS.json after they are logged in and read the account, but sure you don't know if it gets complicated for the average user.

> The way I will build it should cover all these edge case and be rather simple to use (applying sets to the game should not be complicated)

I understand you want to cover all the bases.

> It will not work this way, ItemSet.json gets load only at client launch, if you replace it you must restart the client to load it, and doing so resets it :)

I tested and see the IS,json is not _always_ loaded on client launch. I say not always, because it doesn't if it's the same account as the last, but it has to for different accounts login in. I now you want to cover all the bases :-) . but since I did these tests already I want to at least share them... :-) so to say it wasn't a waste of time I'll paste what I already had written:
-------------------------------------------------------------------------------
I believe ItemSets.json not automatically updated from server on launching LoL client (if we’re the same user as last user logged in)
- LoL client not running
- modify ItemSets.json manually eg. Change name of first Item Set
- launch LoL, go into Item Sets – manually modified name is shown, so server copy was not loaded with old value. As expected, Store doesn’t see the change since it reads \config\Global.

I believe Server copy of ItemSets.json is copied locally only if the user login in is different from last, and I believe Riot assumes local copy of ItemSets.json is always in sync with local copy for the same user since they update ItemsSets.json locally and on the server as changes are saved in Item Set Editor. At that time they also update \config\Champions and \config\Global so the Store is synced as well. I don’t believe they look at the timestamp to determine which file is newer since they assume it’s always in sync.

So, if you create the ItemSets.json files in addition to the \champion and \global and we copy them over the existing ones we will be OK.

There will be a problem if we use different users on our LoL installation. If we load ISE sets and then log in with another user, the second users Item Sets will be loaded, overwriting our ISE sets. Then when we log in again, Riot will restore our old non-ISE sets. 

However, the solution for this – if we use another user on or LoL client or just want to be sure to update our sets on Riot Server for use on another PC – is to copy our new ISE sets, go into LoL Item Set editor, perform a minor change and hit Save to update our sets on Riot Server.
--------------------------------------------------------------------------------------
Anyway, I'll wait for ISE changes. Should we get an email or do we have to check?

Thanks for your efforts.

**Comment by Cadrach** [comment]
I will update this issue once it is live, so you can check here :)

**Comment by OnlyRiven** [comment]
So What is the fix of this problem because I want to download some item sets, but it isn't working...

**Comment by Cadrach** [comment]
Yep sorry, I have started working on it but lacked the time to properly complete it. I will update here once I have had the time to finish it correctly.
